src/mds/SessionMap.h

```diff
@@ -215,20 +216,64 @@ class Session : public RefCountedObject {
     ceph_assert(!info.prealloc_inos.empty());
 
     if (ino) {
-      if (info.prealloc_inos.contains(ino))
+      if (info.prealloc_inos.contains(ino)) {
 	info.prealloc_inos.erase(ino);
-      else
+	if (delegated_inos.contains(ino))
+	  delegated_inos.erase(ino);
+      } else {
 	ino = 0;
+      }
     }
     if (!ino) {
-      ino = info.prealloc_inos.range_start();
-      info.prealloc_inos.erase(ino);
+      /* Grab first prealloc_ino that isn't delegated */
+      for (const auto p: info.prealloc_inos) {
+	for (auto i = p.first ; i < p.first + p.second ; i += 1) {
+	  inodeno_t dstart, dlen;
+	  if (!delegated_inos.contains(i, &dstart, &dlen)) {
+	    ino = i;
+	    info.prealloc_inos.erase(ino);
+	    break;
+	  }
+	  /* skip to end of delegated interval */
+	  i = dstart + dlen - 1;
+	}
+	if (ino)
+	  break;
+      }
     }
+    ceph_assert(ino);
     info.used_inos.insert(ino, 1);
     return ino;
   }
+  void delegate_inos(int want, interval_set<inodeno_t>& newinos) {
+    /* Never take more than half the prealloc_inos */
+    want = std::min((int)info.prealloc_inos.size()/2, want - (int)delegated_inos.size());
```

Comment:
After doing some testing today, I think this std::min may be a problem:

I occasionally see the client flip to doing synchronous creates after a large number of them. It usually corrects itself quickly, so my suspicion is that sometimes the MDS goes to delegate inos to the client, but can't get them because the prealloc pool has run out. Eventually the client ends up starved of inos and has to do a sync create.

I'm now thinking we don't want to clamp this value.